app/lib/multistep/controller.rb

```diff
@@ -0,0 +1,137 @@
+# frozen_string_literal: true
+
+module Multistep
+  module Controller
+    extend ActiveSupport::Concern
+
+    class_methods do
+      def form(value = nil, as: nil) # rubocop:disable Naming/MethodParameterName
+        if value
+          @form = value
+          if as
+            alias_method as, :form
+            helper_method as
+          end
+        else
+          @form
+        end
+      end
+
+      def session_key(value = nil)
+        if value
+          @session_key = value
+        else
+          @session_key ||= @form && @form.model_name.param_key
+        end
+      end
+
+      def params_key(value = nil)
+        if value
+          @params_key = value
+        else
+          @params_key ||= @form && @form.model_name.param_key
+        end
+      end
+
+      def escape_route(&block)
+        if block
+          @escape_route = block
+        else
+          @escape_route || raise("Escape route not defined!")
+        end
+      end
+
+      def result(as: name) # rubocop:disable Naming/MethodParameterName
+        alias_method as, :result
+        helper_method as
+      end
+    end
+
+    included do
+      helper_method :back_link_path
+      before_action :ensure_form_present, only: %i[show update complete]
+      after_action :remove_form, only: :complete
+    end
+
+    attr_reader :result
+
+    def start
+      reset_form
+      redirect_to action: :show, step: form.next_step
+    end
+
+    def show
+      render current_step
+    end
+
+    def update
+      form.assign_attributes(form_params)
+
+      if form.valid?(current_step)
+        form.record_completed_step current_step
+        store_form_in_session
+        redirect_to action: :show, step: step_param(form.next_step)
+      else
+        render current_step
+      end
+    end
+
+    def complete
+      @result = add_participant_form.save!
```

Comment:
Hmm, I would not expect `add_participant_form` mentioned here.

Reply:
Definitely should not! :) Should be `form`